sync_diff_inspector/checkpoints/checkpoints.go

```diff
@@ -0,0 +1,252 @@
+	cp.hp.mu.Lock()
+	defer cp.hp.mu.Unlock()
+	return cp.hp.CurrentSavedNode
+}
+
+func (cp *Checkpoint) Insert(node *Node) {
+	cp.hp.mu.Lock()
+	heap.Push(cp.hp, node)
+	cp.hp.mu.Unlock()
+}
+
+// Len - get the length of the heap
+func (hp *nodeHeap) Len() int { return len(hp.Nodes) }
+
+// Less - determine which is more priority than another
+func (hp *nodeHeap) Less(i, j int) bool {
+	return hp.Nodes[i].IsLess(hp.Nodes[j])
+}
+
+// Swap - implementation of swap for the heap interface
+func (hp *nodeHeap) Swap(i, j int) {
+	hp.Nodes[i], hp.Nodes[j] = hp.Nodes[j], hp.Nodes[i]
+}
+
+// Push - implementation of push for the heap interface
+func (hp *nodeHeap) Push(x interface{}) {
+	hp.Nodes = append(hp.Nodes, x.(*Node))
+}
+
+// Pop - implementation of pop for heap interface
+func (hp *nodeHeap) Pop() (item interface{}) {
+	if len(hp.Nodes) == 0 {
+		return
+	}
+
+	hp.Nodes, item = hp.Nodes[:len(hp.Nodes)-1], hp.Nodes[len(hp.Nodes)-1]
+	return
+}
+
+func (cp *Checkpoint) Init() {
+	hp := &nodeHeap{
+		mu:    &sync.Mutex{},
+		Nodes: make([]*Node, 0),
+		CurrentSavedNode: &Node{
+			ChunkRange: &chunk.Range{
+				Index:   chunk.GetInitChunkID(),
+				IsFirst: true,
+				IsLast:  true,
+			},
+		},
+	}
+	heap.Init(hp)
+	cp.hp = hp
+}
+
+// GetChunkSnapshot get the snapshot of the minimum continuous checked chunk
+func (cp *Checkpoint) GetChunkSnapshot() (cur *Node) {
+	cp.hp.mu.Lock()
+	defer cp.hp.mu.Unlock()
+	for cp.hp.Len() != 0 && cp.hp.CurrentSavedNode.IsAdjacent(cp.hp.Nodes[0]) {
+		cp.hp.CurrentSavedNode = heap.Pop(cp.hp).(*Node)
+		cur = cp.hp.CurrentSavedNode
+	}
+	// wait for next 10s to check
+	return cur
+}
+
+// SaveChunk saves the chunk to file.
+func (cp *Checkpoint) SaveChunk(ctx context.Context, fileName string, cur *Node, reportInfo *report.Report) (*chunk.ChunkID, error) {
+	if cur == nil {
+		return nil, nil
+	}
+
+	savedState := &SavedState{
+		Chunk:  cur,
+		Report: reportInfo,
+	}
+	checkpointData, err := json.Marshal(savedState)
+	if err != nil {
+		log.Warn("fail to save the chunk to the file", zap.Any("chunk index", cur.GetID()), zap.Error(err))
+		return nil, errors.Trace(err)
+	}
+
+	if err = ioutil2.WriteFileAtomic(fileName, checkpointData, config.LocalFilePerm); err != nil {
+		return nil, err
+	}
+	log.Info("save checkpoint",
+		zap.Any("chunk", cur),
+		zap.String("state", cur.GetState()))
+	return cur.GetID(), nil
+}
+
+// LoadChunk loads chunk info from file `chunk`
+func (cp *Checkpoint) LoadChunk(fileName string) (*Node, *report.Report, error) {
+	bytes, err := os.ReadFile(fileName)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	n := &SavedState{}
+	err = json.Unmarshal(bytes, n)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	return n.Chunk, n.Report, nil
+}
+
+// CheckConfig stores the information about the config of the current run.
+// If we restart a sync-diff from a different config, the checkpoint is disable.
+type CheckConfig struct {
+	Table     string `json:"tables"`
```

Comment:
What's this used for?
```suggestion
	Table     string `json:"table"`
```

Reply:
Done. Remove this struct.